[bazel/ci] schedule and run cw310 tests with bazel

drewmacrae opened the pull request:
Marked presently broken tests with a broken tag and recorded issue
Prefer broken tag to other tags that steer CI execution
Copy bitstream to a bazel cache
Run bazel which should grab the [bitstream](https://github.com/lowRISC/opentitan/blob/47526c3e2cb9768fac3c7464eb9ded31dd7ee3aa/rules/bitstreams.bzl#L27) from the cache and run all the
tests we expect to pass
extended timeout to 16 minutes to keep it from getting flaky. (There are
currently more tests than before and the cw310 for our CI isn't a
limiting factor.
Addresses #12445 
Improves but doesn't fix #6965

Signed-off-by: Drew Macrae <drewmacrae@google.com>

milesdai commented:
I saw your comments in #12445 regarding the `broken` tag vs the `failing_{target}` tag. The original intent with keeping separate `failing_verilator` and `failing_cw310` was that these are potentially different classes of problems that we will have to solve going forward, and it might be helpful to be able to filter on one vs the other.

milesdai commented:
Actually, on further thought, you can definitely construct a bazel query to get the intersection between broken tests and the tests for a particular target. @timothytrippel any other thoughts on switching over to the `broken` flag?

milesdai commented on sw/device/tests/BUILD:
```diff
@@ -998,6 +1085,9 @@ opentitan_functest(
 opentitan_functest(
     name = "usbdev_test",
```
I think this test should probably be restricted to the Verilator target: https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/tests/usbdev_test.c;l=95

drewmacrae replied:
Thanks I didn't know that

drewmacrae commented on azure-pipelines.yml:
```diff
@@ -448,18 +448,13 @@ jobs:
-
-      . util/build_consts.sh
-      pytest --version
-      pytest test/systemtest/earlgrey/test_fpga_cw310.py \
```
Yes

timothytrippel commented:
> Actually, on further thought, you can definitely construct a bazel query to get the intersection between broken tests and the tests for a particular target. @timothytrippel any other thoughts on switching over to the `broken` tag from the `-failing_verilator`?

personally, I prefer the more specific "failing_{verilator,cw310}" tags, as opposed to one generic "broken" tag, as sometimes there are cases where a test passes on CW310 but fails on Verilator (this happened to @milesdai and I when we were setting up mask ROM E2E testing). Using the more specific tags would allow us to run as many possible non-failing tests in CI. As opposed to having to not run a CW310 test just because its Verilator counter part is failing.

drewmacrae commented:
> Sometimes there are cases where a test passes on CW310 but fails on Verilator (this happened to @milesdai and I when we were setting up mask ROM E2E testing). Using the more specific tags would allow us to run as many possible non-failing tests in CI. As opposed to having to not run a CW310 test just because its Verilator counter part is failing.

We're tagging the target specific tests already here.
```
cw310 = cw310_params(
    tags = ["broken"],
),
```

So this tag is specific to the target in its placement. The union of tags and the difference is also available in queries and test tag filters. I'm not sure if there's a good way to get the intersection (eg. the broken cw310 tests) of tags with filtering, but we can get it pretty easily with queries.

drewmacrae commented:
The sw build didn't finish in the 16 minutes I allocated to this test. 
4:43- test started
4:46- downloads completed bazel begins to fetch
4:59- test cancelled

drewmacrae commented:
For now we're skipping tests that depend on the mask rom, but we could query for them and run them in a subsequent step after the mask rom is spliced, or delay this step to enable them to run as well.

timothytrippel commented:
> For now we're skipping tests that depend on the mask rom, but we could query for them and run them in a subsequent step after the mask rom is spliced, or delay this step to enable them to run as well.

Quick question, to clarify, do any systemtest tests use the mask ROM bitstream?

drewmacrae commented:
> > For now we're skipping tests that depend on the mask rom, but we could query for them and run them in a subsequent step after the mask rom is spliced, or delay this step to enable them to run as well.
> 
> Quick question, to clarify, do any systemtest tests use the mask ROM bitstream?

Nope, the cw310 systemtests only waits for the build with vivado to complete. It doesn't depend on the splice step so it doesn't do any tests with the mask ROM. (That's therefore beyond the scope for this PR but shouldn't be too big a challenge or too big an impact on CI to include in the future.)

drewmacrae commented:
I was able to reproduce the behavior locally after deleting a splice file and clearing my intermediate caches with `bazel clean --expunge ; bazel sync`. A dummy splice file makes the test_rom tests pass, though the mask rom tests don't pass with a dummy bitstream.

drewmacrae commented:
Rebased to bring in Tim's work on the bitstreams from 5/5 because it looks like it might be relevant.

timothytrippel commented on sw/host/opentitanlib/src/transport/cw310/mod.rs:
```diff
@@ -203,6 +203,12 @@ impl Transport for CW310 {
             usb.pin_set_state(CW310::PIN_SW_STRAP1, true)?;
             usb.pin_set_state(CW310::PIN_SW_STRAP2, true)?;
+            usb.pin_set_state(CW310::PIN_TAP_STRAP1, true)?;
+            usb.pin_set_state(CW310::PIN_SRST, false)?;
+            std::thread::sleep(fpga_program.rom_reset_pulse);
+            usb.pin_set_state(CW310::PIN_SRST, true)?;
```
I don't think this reset sequence is needed here is it? since bootstrap will issue this reset sequence [here](https://cs.opensource.google/opentitan/opentitan/+/master:sw/host/opentitanlib/src/bootstrap/mod.rs;drc=a4008df8c696af2a514e9964a9e5996c21f4de0e;l=168)

drewmacrae replied:
For a chip that has 2 reset pins and 5 strapping pins, I'm not confident in that bootstrapping, but I could try it without a reset and if it works and I need to restart CI I'd be happy to push the change.

drewmacrae replied:
Yep, you're right

timothytrippel commented on sw/host/opentitanlib/src/transport/cw310/mod.rs:
```diff
@@ -203,6 +203,12 @@ impl Transport for CW310 {
             usb.pin_set_state(CW310::PIN_SW_STRAP0, true)?;
             usb.pin_set_state(CW310::PIN_SW_STRAP1, true)?;
             usb.pin_set_state(CW310::PIN_SW_STRAP2, true)?;
+            usb.pin_set_state(CW310::PIN_TAP_STRAP1, true)?;
```
i think whats most important here is setting sensible defaults states for the tap straps to match [here](https://github.com/lowRISC/opentitan/pull/12633#discussion_r871938097), which this does. nice catch!

drewmacrae replied:
I think it's actually above where we clear PIN_TAP_STRAP0 so the SPI bus gets enabled when we also clear PIN_TAP_STRAP1, the rest is just to bring this in-line with the tested behavior until this is the tested behavior.

timothytrippel commented on sw/host/opentitanlib/src/transport/cw310/mod.rs:
```diff
@@ -203,6 +203,12 @@ impl Transport for CW310 {
             usb.pin_set_state(CW310::PIN_SW_STRAP0, true)?;
             usb.pin_set_state(CW310::PIN_SW_STRAP1, true)?;
             usb.pin_set_state(CW310::PIN_SW_STRAP2, true)?;
```
these setting of defaults for the various strap pins (and the ones below) are only run if the fpga is re-programmed with a new bitstream, if it is determined that a new bitstream is not required (on line 196 above) we could return early without ever setting sensible defaults. Perhaps its worth moving all of these default pin state assignments to a separated function like the `init_direction()` function above (you could call it `init_pin_states` or something like that, and call this function  in the CW310 transport constructor, that way we start from a clean known state everytime.

timothytrippel commented:
Given this PR has gone through quite a bit a of churn, I'm cool with fixing these opentitantool pin inits in  a follow up PR if this passes CI. Theoretically, what you have done should make things work, they just aren't quite all necessary / in the right place for long-term stability. All good tho, thanks for the quick fix (fingers crossed this finally passes) !

drewmacrae commented on sw/host/opentitanlib/src/transport/cw310/mod.rs:
```diff
@@ -76,7 +76,7 @@ impl CW310 {
         device.pin_set_output(Self::PIN_TRST, true)?;
         device.pin_set_output(Self::PIN_SRST, true)?;
-        device.pin_set_output(Self::PIN_TAP_STRAP0, true)?;
+        device.pin_set_output(Self::PIN_TAP_STRAP0, false)?;
```
Ok, so I'm familiar with GPIO API's that don't do it this way... I will fix that.

timothytrippel commented:
As discussed offline, the opentitantool fixes have been factored out into a separate PR: https://github.com/lowRISC/opentitan/pull/12723

Feel free to rebase on this PR when you get a chance.

drewmacrae commented:
rebased and pushed to drop cherry-pick

timothytrippel commented:
are we good to merge this?

drewmacrae commented:
There's a bit of flakiness I'm trying to protect the CI from, ~1% of tests run appear to be failing, but the failures aren't independent and sometimes appear in pairs, which makes it a little harder to rerun to accept. We'll log the flakiness, I just don't want it to impact peoples PRs if I can help it.

rswarbrick commented:
How odd! Were we seeing the equivalent flakiness with the pytest framework before? I wouldn't be that surprised if we sometimes had to re-run stuff, but I don't remember doing it myself recently. If not, this seems like a regression and it would be good to understand what's going on before making things worse.

drewmacrae commented:
The flakiness appears to be specific to some OpenTitan-tool functionality that isn't present in the pytest framework. IMO there's enough built in mechanisms to work around and log the flakiness, I've rebased and added the relevant tags and flags. 

It's a regression and we can reproduce it pretty easily with bazel and the associated software that this PR is going to help us continue to maintain and improve.
```
bazel test --test_tag_filters=cw310,-broken --build_tests_only //... 
--runs_per_test=10 --runs_per_test_detects_flakes
```

drewmacrae commented:
I think the opentitan-tool flakiness should be resolved here. It will take some time to test it on my machine.
